src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -581,11 +581,27 @@ void ProcessMarkedTypesWithInterfaces ()
 			foreach ((var type, var scope) in typesWithInterfaces) {
 				// Exception, types that have not been flagged as instantiated yet.  These types may not need their interfaces even if the
 				// interface type is marked
-				if (!Annotations.IsInstantiated (type) && !Annotations.IsRelevantToVariantCasting (type))
+				// UnusedInterfaces optimization is turned off mark all interface implementations
+				bool unusedInterfacesOptimizationEnabled = Context.IsOptimizationEnabled (CodeOptimizations.UnusedInterfaces, type);
+				if (!Annotations.IsInstantiated (type) && !Annotations.IsRelevantToVariantCasting (type) &&
+					unusedInterfacesOptimizationEnabled)
 					continue;
 
-				using (ScopeStack.PushScope (scope))
+				using (ScopeStack.PushScope (scope)) {
 					MarkInterfaceImplementations (type);
+
+					if (unusedInterfacesOptimizationEnabled)
+						continue;
+
+					// If the optimization is disabled, make sure to mark all methods which implement interfaces
```

Comment:
You might be right - I must admit that I still don't fully grasp the exact order in which things happen in this case.
I'm pretty sure the comment was there because that was what I wanted to do - but then I realized that I don't need that actually since ProcessOverride will do it for me. So I simplified the code. It's possible it can be entirely removed.

What the code does currently is that it will mark methods which are overrides/impls of methods from non-link assemblies (in library mode that's basically anything outside of the current assembly). I'm not sure if we can remove that fully - it depends if in library mode we mark all of the non-link assemblies, I'm not sure we do.

Reply:
I'm not 100% sure what we currently do for "library mode" since we never really came up with a design for it, but I think you're right that we don't mark assemblies other than the library. That's also what would make sense to me.

So I guess `ProcessOverride` won't be called for the case of a non-virtual method implementing an interface method defined in a referenced assembly - then I can see why we would need this logic. I wonder if there's a way we could simplify this, since it's very similar to `IsVirtualNeededByTypeDueToPreservedScope`. It feels like we want to do substantially the same thing for virtual method overrides and for interface method implementations.

Maybe that could be generalized to `IsMethodNeededByTypeDueToPreservedScope` so that it works for non-virtual methods, and methods implementing interfaces.